server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -49,54 +53,77 @@ import           Hasura.Prelude
+    -- ^ The hasura to remote schema field mapping.
     , rrArguments     :: !ArgsMap
+    -- ^ See '_fArguments'.
     , rrSelSet        :: !SelSet
+    -- ^ Selection set of 'rrAlias' (may be empty)
     , rrRelFieldPath  :: !RelFieldPath
+    -- ^ Splice path.
     , rrAlias         :: !G.Alias
+    -- ^ A possible alias (provided by client) for `rtrName .
+    -- rmfRemoteRelationship . rrRemoteField`, otherwise this will be defaulted
+    -- to that value.
     , rrAliasIndex    :: !Int
+    -- ^ index of selection set at 'rrRelFieldPath' at which this remote field should be spliced.
+    -- This is morally the last component of 'rrRelFieldPath'.
     , rrPhantomFields :: ![Text]
+    -- ^ Hasura fields which are not in the selection set, but are required as
+    -- parameters to satisfy the remote join.
     }
   deriving (Show)
 
+-- | A path of ordered keys (suitable for 'OJ.insert') allowing us to splice a
+-- remote result at a nested location.
 newtype RelFieldPath =
   RelFieldPath {
     unRelFieldPath :: (Seq.Seq (Int, G.Alias))
   } deriving (Show, Monoid, Semigroup, Eq)
 
-data InsertPath =
-  InsertPath
-    { ipFields :: !(Seq.Seq Text)
-    , ipIndex  :: !(Maybe ArrayIndex)
-    }
-  deriving (Show, Eq)
-
-newtype ArrayIndex =
-  ArrayIndex Int
-  deriving (Show, Eq, Ord)
-
--- Rebuild the field with remote relationships removed, and paths that
--- point back to them.
-rebuildFieldStrippingRemoteRels ::
-     VQ.Field -> Maybe (VQ.Field, NonEmpty RemoteRelField)
+-- | Rebuild the query tree, returning a new tree with remote relationships
+-- removed, along with a list of paths that point back to them.
+--
+-- 'Nothing' means there were no remote fields.
+--
+-- Output here is eventually assembled into a 'QExecPlan' 'Tree' in
+-- 'getExecPlan', and finally executed in 'runGQ'.
+rebuildFieldStrippingRemoteRels
+  :: VQ.Field 
+  -> Maybe (VQ.Field, NonEmpty RemoteRelField)
+  -- ^ NOTE: the ordering of the _fSelSet fields in the result 'VQ.Field' seems not to matter at 
+  -- all for correctness here (based on experimentation)... I haven't puzzled out why yet.
 rebuildFieldStrippingRemoteRels =
   traverse NE.nonEmpty . flip runState mempty . rebuild 0 mempty
   where
+    rebuild :: Int -> RelFieldPath -> Field -> State [RemoteRelField] Field
     rebuild idx0 parentPath field0 = do
-      selSetEithers <-
-        for (zip [0..] (toList (_fSelSet field0))) $
-          \(idx, subfield) ->
+      forMOf fSelSet field0 $ \ss0 ->
+        fmap join $ flip Seq.traverseWithIndex ss0 $
+          \idx subfield ->
              case _fRemoteRel subfield of
-               Nothing -> fmap Right (rebuild idx thisPath subfield)
+               Nothing -> Seq.singleton <$> rebuild idx thisPath subfield
                Just remoteField -> do
+
                  modify (remoteRelField :)
-                 pure (Left remoteField)
+                 -- NOTE: this may result in redundant fields in the SELECT, but this seems fine.
```

Comment:
Not sure I understand what you mean by `redundant fields in the SELECT`?

Reply:
Sorry I meant to trace this again, but I saw a select with json aggregation with duplicated keys which postgres just ignores/unions in some way.